### PR TITLE
fix(blob)!: update the blob.Proof to return a proof to the data root

### PR DESCRIPTION
**Pull request by rach-id**
Closes https://github.com/celestiaorg/celestia-node/issues/3582

### CODE REVIEW

File: blob/blob.go [hunk 1]
```diff
@@ -18,42 +21,107 @@ import (
...
+				return ErrInvalidProof
+			}
+		}
 	}
 	return nil
 }
 
+// Verify takes a set of shares, a namespace and a data root, and verifies if the
+// provided shares are committed to by the data root.
+func (p *Proof) Verify(rawShares [][]byte, namespace share.Namespace, dataRoot []byte) (bool, error) {
```
Comment (rach-id) [inline blob/blob.go:88]: We will also add another method that will verify this proof but using the subtree roots

Comment (rach-id) [reply blob/blob.go:88]: https://github.com/celestiaorg/celestia-node/issues/3611

File: blob/blob.go [hunk 2]
```diff
@@ -16,44 +18,104 @@ import (
 	"github.com/celestiaorg/celestia-node/share"
 )
 
+//nolint:unused
 var errEmptyShares = errors.New("empty shares")
```
Comment (Wondertan) [inline blob/blob.go:22]: Let's remove it if its unused

File: blob/parser.go [hunk 3]
```diff
@@ -21,6 +21,8 @@ type parser struct {
 
 // set tries to find the first blob's share by skipping padding shares and
 // sets the metadata of the blob(index and length)
+//
+//nolint:unused
```
Comment (Wondertan) [inline blob/parser.go:25]: @vgonkivs, should we keep this code if it's unused now?

File: api/docgen/exampledata/blobProof.json [hunk 4]
```diff
@@ -1,31 +1,53 @@
...
-    "nodes": [
-      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAMJ/xGlNMdEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAwn/EaU0x0UTO9HUGKjyjcv5U2gHeSjJ8S1rftqv6k8kxlVWW8e/7",
-      "/////////////////////////////////////////////////////////////////////////////wexh4khLQ9HQ2X6nh9wU5B+m6r+LWwPTEDTa5/CosDF"
-    ],
-    "is_max_namespace_ignored": true
+    "start_row": 1,
+    "end_row": 3
   }
-]
+}
```
Comment (Wondertan) [inline api/docgen/exampledata/blobProof.json:53]: new line

File: share/eds/edstest/testing.go [hunk 5]
```diff
@@ -138,13 +139,22 @@ func createTestBlobTransaction(
 	ns := namespace.RandomBlobNamespace()
 	msg, blob := blobfactory.RandMsgPayForBlobsWithNamespaceAndSigner(addr.String(), ns, size)
 	require.NoError(t, err)
+	cTx := BuildCoreTx(t, signer, msg, blob)
+	return ns, msg, blob, cTx
+}
 
+func BuildCoreTx(
```
Comment (Wondertan) [inline share/eds/edstest/testing.go:146]: A little godoc description wouldn't hurt

File: blob/service_test.go [hunk 6]
```diff
@@ -287,7 +281,9 @@ func TestBlobService_Get(t *testing.T) {
 				require.ErrorIs(t, err, ErrInvalidProof)
 				included, ok := res.(bool)
 				require.True(t, ok)
-				require.True(t, included)
+				// This is supposed to be false not true.
+				// we're querying blob[1]'s proof and verifying it against blob[0], no?
+				require.False(t, included)
```
Comment (Wondertan) [inline blob/service_test.go:286]: cc @vgonkivs

File: blob/blob.go (hunk 2 above)
Comment (rach-id) [reply blob/blob.go:22]: want to confirm with @vgonkivs  first

File: blob/parser.go (hunk 3 above)
Comment (Wondertan) [reply blob/parser.go:25]: Actually, we are likely gonna use it later on, but on the server side. The only way for the server to serve blobs with proofs is by parsing the whole namespace/row and this code is gonna be useful there, so we might keep it for now

File: blob/service.go [hunk 7]
```diff
@@ -351,123 +362,187 @@ func (s *Service) retrieve(
...
-			err = ErrBlobNotFound
+	// get all the shares of the rows containing the namespace
+	_, getSharesSpan := tracer.Start(ctx, "get-all-shares-in-namespace")
+	// store the ODS shares of the rows containing the blob
+	rowsShares := make([]share.Share, 0, (exclusiveNamespaceEndRowIndex-inclusiveNamespaceStartRowIndex)*squareSize)
+	// store the EDS shares of the rows containing the blob
+	rowsWithParityShares := make([][]shares.Share, exclusiveNamespaceEndRowIndex-inclusiveNamespaceStartRowIndex)
+	for rowIndex := inclusiveNamespaceStartRowIndex; rowIndex < exclusiveNamespaceEndRowIndex; rowIndex++ {
+		for colIndex := 0; colIndex < squareSize*2; colIndex++ {
+			share, err := s.shareGetter.GetShare(ctx, header, rowIndex, colIndex)
```
Comment (vgonkivs) [inline blob/service.go:397]: having a blob that consists of 256 shares(or in case of GetAll request), you will make 256 independent network calls in case of the light node. Can we rework it by replacing with GetEDS instead?

File: blob/service.go [hunk 8]
```diff
@@ -351,123 +362,187 @@ func (s *Service) retrieve(
...
+		// skip if it's a padding share
+		isPadding, err := currentShareApp.IsPadding()
 		if err != nil {
 			return nil, nil, err
 		}
+		if isPadding {
+			currentShareIndex++
+			continue
+		}
+		isCompactShare, err := currentShareApp.IsCompactShare()
```
Comment (vgonkivs) [inline blob/service.go:435]: Is this something new? as I'm not aware about `compactShares`?

**Review by vgonkivs — COMMENTED** [review]
Left some comments. Generally speaking the changes look very good but I still need time to understand `blob.Index `calculation. 


P.S. I believe [this](https://github.com/celestiaorg/celestia-node/issues/2289) issue is not relevant anymore: blob retrieval will download **all** shares anyway, so the downloaded commitment will differ from the passed one, in case smn wants to pass a commitment without 1 share.

File: blob/service.go (hunk 8 above)
Comment (rach-id) [reply blob/service.go:435]: those are transaction shares: https://celestiaorg.github.io/celestia-app/specs/shares.html#transaction-shares
Could be any Celestia transaction.

File: blob/blob.go (hunk 2 above)
Comment (rach-id) [reply blob/blob.go]: afb2293718e9d4413e7761dfe2c697cb83726a77

File: api/docgen/exampledata/blobProof.json (hunk 4 above)
Comment (rach-id) [reply api/docgen/exampledata/blobProof.json]: 052a79751d133df92da76bf51317e9764b4762cd

File: share/eds/edstest/testing.go (hunk 5 above)
Comment (rach-id) [reply share/eds/edstest/testing.go:149]: a006f94a188ece637a3be15cdd8d14be8876fd2a

File: blob/service_test.go (hunk 6 above)
Comment (rach-id) [reply blob/service_test.go]: 698d90671df4045c5c4ce1bd59713d4826434206

File: blob/blob.go [hunk 9]
```diff
@@ -16,44 +18,104 @@ import (
...
-		for i, node := range pNodes {
-			if !bytes.Equal(node, inputNodes[i]) {
+func (p *Proof) equal(input *Proof) error {
+	// compare the NMT proofs
+	for i, proof := range p.ShareToRowRootProof {
+		if proof.Start != input.ShareToRowRootProof[i].Start {
+			// TODO(@rach-id): should we define specific errors for each case? that's better
+			// to know which part is not equal.
+			// Also, this error should be ErrProofNotEqual or similar, and not ErrInvalidProof
+			return ErrInvalidProof
```
[the remark this answers is not in the document]
Comment (rach-id) [reply blob/blob.go]: ccb5799c7cd4575d78735f3f6244da44decfa9e2

File: blob/service.go [hunk 10]
```diff
@@ -315,15 +318,23 @@ func (s *Service) Included(
...
 		return false, nil
 	default:
 		return false, err
 	}
-	return true, resProof.equal(*proof)
+	appShares, err := BlobsToShares(blob)
+	if err != nil {
+		return false, err
+	}
+	getter, err := s.headerGetter(ctx, height)
```
[the remark this answers is not in the document]
Comment (rach-id) [reply blob/service.go]: 952b4add6b2560cd2e2c77b4a1b32516816ea445

File: blob/service.go (hunk 7 above)
Comment (rach-id) [reply blob/service.go:393]: We should then define a threshold that we're over it, we should just get the whole EDS. Otherwise, we will be querying the whole EDS for any single blob even if it's just 1 row...

Comment (Wondertan) [reply blob/service.go:393]: Fetching the whole EDS will take more bandwidth, but it will be substantially faster. If we want to keep assuming the LN cannot afford that even for a few months from now, then we can do it by shares, but it will be slower and lead to more load on FN/BN

Note that this is based on our understanding and intuition of how the current protocol works, and we don't have a proper benchmark to prove it.

Comment (Wondertan) [reply blob/service.go:393]: I assume there is no way to do it over `SharesByNamespace`, right?

Comment (rach-id) [reply blob/service.go:393]: `SharesByNamespace` doesn't provide enough information to generate the proof.

We could set a limit to 100 shares, for example. If the queried shares are more, then we query the EDS directly, otherwise we query the shares separately. What do you think?

Later on, we can improve on the value based on the blobs sizes that are submitted to Celestia and how often their proofs are queried.

**Comment by rach-id** [comment]
> P.S. I believe https://github.com/celestiaorg/celestia-node/issues/2289 issue is not relevant anymore: blob retrieval will download all shares anyway, so the downloaded commitment will differ from the passed one, in case smn wants to pass a commitment without 1 share.

Agreed. And in the case of a rollup referencing data using the share commitment, we can document somewhere the possibilities so that they're aware of this.

File: blob/service.go (hunk 7 above)
Comment (vgonkivs) [reply blob/service.go:393]: I'd any advocate for `GetEDS`. Assuming you have an EDS with blobs[b1,b2,b3,b4]. So in the worst case, you'd have to download the whole EDS to get b4, in the best case - only n shares of the first blob. I agree that GetEDS will take a lot more bandwidth in case of the best case, but imo it will be much much faster than the worst case. Also, GetEDS fits more for the GetAll request, since you will anyway have to fetch al the shares under the same namespace. We should also think about users that are actively using `GetAll`: getting shares one by one might take some time, so there might cases when the request will be timeouted.
Setting a threshold for N shares IMO is not the case(and the such logic will be harder to understand and maintain, which could lead a ton of bugs. As Blob Module is pretty hot and popular within the node, it will be painful to fix those bugs), since you don't know whether the share belongs to the blob that you are looking for(you will detect it once download all shares of the blob a create a commitment).

I personally think that should sacrifice the bandwidth in favour of speed.

Comment (rach-id) [reply blob/service.go:393]: sure, as you prefer. This means that we don't need to improve the `GetShareProof` to only get the needed shares :D I'll update the implemenetation

Comment (rach-id) [reply blob/service.go]: [2d209ad](https://github.com/celestiaorg/celestia-node/pull/3610/commits/2d209adef1e5cd662efe2994b60236a95ca76d4e)

**Comment by vgonkivs** [comment]
Agreed in DM with @rach-id that the proof Verification logic should additionally check commitment: whether it belongs to the blob or not.

**Comment by rach-id** [comment]
> Agreed in DM with @rach-id that the proof Verification logic should additionally check commitment: whether it belongs to the blob or not.

ae65e8d

File: blob/service.go [hunk 11]
```diff
@@ -347,127 +354,204 @@ func (s *Service) retrieve(
...
+	}
+	if inclusiveNamespaceStartRowIndex == -1 {
+		return nil, nil, ErrBlobNotFound
+	}
+
+	// end exclusive index of the row root containing the namespace
+	exclusiveNamespaceEndRowIndex := inclusiveNamespaceStartRowIndex
+	for i, row := range header.DAH.RowRoots[inclusiveNamespaceStartRowIndex:] {
+		if namespace.IsOutsideRange(row, row) {
+			exclusiveNamespaceEndRowIndex = inclusiveNamespaceStartRowIndex + i
```
[the remark this answers is not in the document]
Comment (rach-id) [reply blob/service.go]: [e473057](https://github.com/celestiaorg/celestia-node/pull/3610/commits/e473057d5f0e3a9486dce557a5565c7e7962c809)

File: blob/service.go [hunk 12]
```diff
@@ -347,127 +354,204 @@ func (s *Service) retrieve(
...
+			if endShareIndex > len(rowsShares) {
+				// this blob spans to the next row which has a namespace > requested namespace.
+				// this means that we can stop.
+				return nil, nil, ErrBlobNotFound
+			}
+			// convert the blob shares to app shares
+			blobShares := rowsShares[currentShareIndex:endShareIndex]
+			appBlobShares, err := toAppShares(blobShares...)
+			if err != nil {
+				return nil, nil, err
```
Comment (vgonkivs) [inline blob/service.go:466]: Can we just create a blob here? We already know the exact length of it and get all shares, so, let's just call `NewBlob`. It will simplify logic since you don't need using parser anymore.

File: blob/service.go [hunk 13]
```diff
@@ -347,127 +354,204 @@ func (s *Service) retrieve(
...
 
-			shrs, isComplete = sharesParser.addShares(appShares)
-			// move to the next row if the blob is incomplete
+			// parse the blob
+			sharesParser.length = blobLen
+			_, isComplete := sharesParser.addShares(appBlobShares)
 			if !isComplete {
-				appShares = nil
-				break
+				return nil, nil, fmt.Errorf("expected the shares to construct a full blob")
```
Comment (vgonkivs) [inline blob/service.go:473]: Can be removed. See comment above.

File: blob/service.go [hunk 14]
```diff
@@ -347,127 +354,204 @@ func (s *Service) retrieve(
...
 			}
-			// otherwise construct blob
 			blob, err := sharesParser.parse()
 			if err != nil {
 				return nil, nil, err
 			}
```
Comment (vgonkivs) [inline blob/service.go:478]: can be also removed

File: blob/service.go [hunk 15]
```diff
@@ -347,127 +354,204 @@ func (s *Service) retrieve(
...
-					appShares = shrs
-				}
+			endShareIndex := currentShareIndex + blobLen
+			if endShareIndex > len(rowsShares) {
+				// this blob spans to the next row which has a namespace > requested namespace.
+				// this means that we can stop.
+				return nil, nil, ErrBlobNotFound
+			}
+			// convert the blob shares to app shares
+			blobShares := rowsShares[currentShareIndex:endShareIndex]
```
Comment (vgonkivs) [inline blob/service.go:463]: should be exclusive, no? since `endShareIndex := currentShareIndex + blobLen`

File: blob/service.go [hunk 16]
```diff
@@ -347,127 +354,204 @@ func (s *Service) retrieve(
...
+			sharesFromCurrentRowStart := currentShareIndex % squareSize
+			// setting the index manually since we didn't use the parser.set() method
+			blob.index = sharesFromSquareStartToNsStart +
+				sharesFromNsStartToCurrentRow +
+				sharesFromCurrentRowStart
+
+			if blob.Namespace().Equals(namespace) && sharesParser.verify(blob) {
+				// now that we found the requested blob, we will create
+				// its inclusion proof.
+				inclusiveBlobStartRowIndex := inclusiveNamespaceStartRowIndex + currentShareIndex/squareSize
```
Comment (vgonkivs) [inline blob/service.go:498]: Do we need this calculation after we've calculated the `blob.Index`? can't it be calculated this way: `blob.Index/(squareSize*2)`?

File: blob/service.go [hunk 17]
```diff
@@ -347,127 +354,204 @@ func (s *Service) retrieve(
...
-				"namespace: %s detected at %d: %w", namespace.String(), height, err)
-			log.Error(err)
-		}
-	}
-	return nil, nil, err
+// proveRowRootsToDataRoot creates a set of binary merkle proofs for all the
+// roots defined by the range [start, end).
+func proveRowRootsToDataRoot(roots [][]byte, start, end int) []*merkle.Proof {
+	_, proofs := merkle.ProofsFromByteSlices(roots)
+	return proofs[start:end]
```
Comment (vgonkivs) [inline blob/service.go:557]: Can we move this function to the blob.go, where Proof is defined?

File: blob/service.go [hunk 18]
```diff
@@ -347,127 +354,204 @@ func (s *Service) retrieve(
...
+			// setting the index manually since we didn't use the parser.set() method
+			blob.index = sharesFromSquareStartToNsStart +
+				sharesFromNsStartToCurrentRow +
+				sharesFromCurrentRowStart
+
+			if blob.Namespace().Equals(namespace) && sharesParser.verify(blob) {
+				// now that we found the requested blob, we will create
+				// its inclusion proof.
+				inclusiveBlobStartRowIndex := inclusiveNamespaceStartRowIndex + currentShareIndex/squareSize
+				exclusiveBlobEndRowIndex := inclusiveNamespaceStartRowIndex + endShareIndex/squareSize
```
Comment (vgonkivs) [inline blob/service.go:499]: I would not call it exclusive as you include it in every range:
`range header.DAH.RowRoots[inclusiveBlobStartRowIndex:exclusiveBlobEndRowIndex]`

**Review by vgonkivs — COMMENTED** [review]
Made one more review round 🙂. During it I noted all the variable that were used in PR with brief explanation 🤣. I have an impression that it will cause a lot confusion in future when we revisit this code. I want to propose to rename them now: since we are dealing with ods/eds shares, you include those into namings. It will definitely simplify the understanding.

File: blob/service.go [hunk 19]
```diff
@@ -347,127 +354,204 @@ func (s *Service) retrieve(
...
+	_, getSharesSpan := tracer.Start(ctx, "get-all-shares-in-namespace")
+	// store the ODS shares of the rows containing the blob
+	rowsShares := make([]share.Share, 0, (exclusiveNamespaceEndRowIndex-inclusiveNamespaceStartRowIndex)*squareSize)
+	// store the EDS shares of the rows containing the blob
+	rowsWithParityShares := make([][]shares.Share, exclusiveNamespaceEndRowIndex-inclusiveNamespaceStartRowIndex)
+
+	for rowIndex := inclusiveNamespaceStartRowIndex; rowIndex < exclusiveNamespaceEndRowIndex; rowIndex++ {
+		for colIndex := 0; colIndex < squareSize*2; colIndex++ {
+			share := eds.GetCell(uint(rowIndex), uint(colIndex))
+			if colIndex < squareSize {
```
[the remark this answers is not in the document]
Comment (rach-id) [reply blob/service.go]: [e5677a8](https://github.com/celestiaorg/celestia-node/pull/3610/commits/e5677a8d498f140555f7213e050ee972e97de866)

File: blob/service.go [hunk 20]
```diff
@@ -347,127 +354,204 @@ func (s *Service) retrieve(
...
 
-	// collect shares for the requested namespace
-	namespacedShares, err := s.shareGetter.GetSharesByNamespace(getCtx, header, namespace)
-	if err != nil {
-		if errors.Is(err, share.ErrNotFound) {
-			err = ErrBlobNotFound
+	// get all the shares of the rows containing the namespace
+	_, getSharesSpan := tracer.Start(ctx, "get-all-shares-in-namespace")
+	// store the ODS shares of the rows containing the blob
+	rowsShares := make([]share.Share, 0, (exclusiveNamespaceEndRowIndex-inclusiveNamespaceStartRowIndex)*squareSize)
```
[the remark this answers is not in the document]
Comment (rach-id) [reply blob/service.go]: [d7bb13b](https://github.com/celestiaorg/celestia-node/pull/3610/commits/d7bb13b4ff89bd00fab59635e8e49e0bb5f20bb7)

File: blob/service.go [hunk 21]
```diff
@@ -347,127 +354,204 @@ func (s *Service) retrieve(
...
+	// store the EDS shares of the rows containing the blob
+	rowsWithParityShares := make([][]shares.Share, exclusiveNamespaceEndRowIndex-inclusiveNamespaceStartRowIndex)
```
[the remark this answers is not in the document]
Comment (rach-id) [reply blob/service.go]: [11bfc48](https://github.com/celestiaorg/celestia-node/pull/3610/commits/11bfc485c48197427e951683d35e400ccea6416f)

File: blob/service.go (hunk 12 above)
Comment (rach-id) [reply blob/service.go:459]: I keep this logic because we call `sharesParser.verify()` which can have custom verification function and it's used by `getAll` etc...

File: blob/service.go (hunk 13 above)
Comment (rach-id) [reply blob/service.go:466]: https://github.com/celestiaorg/celestia-node/pull/3610#discussion_r1702092515

File: blob/service.go (hunk 14 above)
Comment (rach-id) [reply blob/service.go:471]: https://github.com/celestiaorg/celestia-node/pull/3610#discussion_r1702092515

File: blob/service.go (hunk 15 above)
Comment (rach-id) [reply blob/service.go]: yes, the `endShareIndex` is exclusive. Renamed in: 500f0b8

For example, we have the following row:

0 1 2 3 4 5 6 7

currentShareIndex := 2
blobLen := 3

since `currentShareIndex` is the first share of the blob, the end share of the blob is `4` : `{2, 3, 4}`

2 + 3 = 5 which is exclusive

File: blob/service.go (hunk 16 above)
Comment (rach-id) [reply blob/service.go:491]: yes, want me to update?

```suggestion
				inclusiveBlobStartRowIndex := blob.index / (squareSize * 2)
```

File: blob/service.go (hunk 17 above)
Comment (rach-id) [reply blob/service.go]: [11d425c](https://github.com/celestiaorg/celestia-node/pull/3610/commits/11d425c3fe02293c3229f14bad985d6a5a075f56)

File: blob/service.go (hunk 18 above)
Comment (rach-id) [reply blob/service.go]: the ranges when slicing go slices are exclusive, so it's exclusive, no?

File: blob/service.go (hunk 12 above)
Comment (vgonkivs) [reply blob/service.go:458]: we can rework retrieve by applying `verifyFn() `. But it's up to you. we can do it later

File: blob/service.go (hunk 15 above)
Comment (vgonkivs) [reply blob/service.go]: As it exclusive, shouldn't it be:
`blobShares := rowsShares[currentShareIndex:endShareIndex-1]`
currentShareIndex := 2
blobLen := 3
2+3 =5 
But the actual end share is 4

File: blob/service.go (hunk 16 above)
Comment (vgonkivs) [reply blob/service.go]: yes, please

File: blob/service.go (hunk 18 above)
Comment (vgonkivs) [reply blob/service.go]: true. But you should range over 
header.DAH.RowRoots[inclusiveBlobStartRowIndex:exclusiveBlobEndRowIndex-1] then, no?

File: blob/service.go (hunk 12 above)
Comment (rach-id) [reply blob/service.go:458]: yes, but didn't want to remove the share parser in this PR and have to change unrelated code too. Let's leave it for a separate PR

File: blob/service.go (hunk 15 above)
Comment (rach-id) [reply blob/service.go]: for example: `slice := []int{0, 1, 2, 3, 4}`
`slice[0:2] =  {0, 1}`.

The end range is not included in the go slice. Go slicing is end exclusive

File: blob/service.go (hunk 18 above)
Comment (rach-id) [reply blob/service.go]: https://github.com/celestiaorg/celestia-node/pull/3610#discussion_r1703865887

File: blob/blob.go [hunk 22]
```diff
@@ -16,42 +19,70 @@ import (
...
-	if p.Len() != input.Len() {
-		return ErrInvalidProof
+// Verify takes a blob and a data root and verifies if the
+// provided blob was committed to by the data root.
+func (p *Proof) Verify(blob *Blob, dataRoot []byte) (bool, error) {
+	blobCommitment, err := types.CreateCommitment(ToAppBlobs(blob)[0])
+	if err != nil {
+		return false, err
+	}
+	if !blob.Commitment.Equal(blobCommitment) {
```
Comment (renaynay) [inline blob/blob.go:42]: Here we are checking if the blob commitment we generate using the given blob matches the commitment inside the given blob? 

What does this have to do with the proof then (re: error message)? wouldn't it just be a malformed blob?

File: blob/blob.go [hunk 23]
```diff
@@ -16,42 +19,70 @@ import (
...
+	RowProof coretypes.RowProof
+}
 
-// equal is a temporary method that compares two proofs.
-// should be removed in BlobService V1.
-func (p Proof) equal(input Proof) error {
-	if p.Len() != input.Len() {
-		return ErrInvalidProof
+// Verify takes a blob and a data root and verifies if the
+// provided blob was committed to by the data root.
```
Comment (renaynay) [inline blob/blob.go:36]:
```suggestion
// Verify takes a blob and a data root and verifies if the
// provided blob was committed to the given data root.
```

**Review by renaynay — COMMENTED** [review]
very difficult to verify the parsing logic here, so could benefit from more tests with malformed proofs + blobs

File: blob/blob.go [hunk 24]
```diff
@@ -16,42 +18,63 @@ import (
...
-
-func (p Proof) Len() int { return len(p) }
+// Proof constructs the proof of a blob to the data root.
+type Proof struct {
+	// ShareToRowRootProof the proofs of the shares to the row roots they belong to.
+	// If the blob spans across multiple rows, then this will contain multiple proofs.
+	ShareToRowRootProof []*tmproto.NMTProof
+	// RowProof the proofs of the row roots containing the blob shares
+	// to the data root.
+	RowProof coretypes.RowProof
```
[the remark this answers is not in the document]
Comment (rach-id) [reply blob/blob.go]: [5b054bc](https://github.com/celestiaorg/celestia-node/pull/3610/commits/5b054bc56d12437886dbe2a9f681eb0d57c5c853)

File: blob/helper.go [hunk 25]
```diff
@@ -39,7 +39,7 @@ func BlobsToShares(blobs ...*Blob) ([]share.Share, error) {
 
 // ToAppBlobs converts node's blob type to the blob type from celestia-app.
 func ToAppBlobs(blobs ...*Blob) []*apptypes.Blob {
-	appBlobs := make([]*apptypes.Blob, 0, len(blobs))
+	appBlobs := make([]*apptypes.Blob, len(blobs))
```
[the remark this answers is not in the document]
Comment (rach-id) [reply blob/helper.go:42]: was buggy before

File: blob/blob.go (hunk 22 above)
Comment (rach-id) [reply blob/blob.go:42]: that's a malformed blob, yes. should I create a new error type?

Comment (renaynay) [reply blob/blob.go:42]: Not sure if a whole new error type is necessary unless we have multiple places where a malformed blob can be detected. regardless, we need the err msg to be accurate

File: blob/blob.go [hunk 26]
```diff
@@ -16,42 +19,70 @@ import (
...
+	ns := append([]byte{namespace.Version()}, namespace.ID()...)
+	cursor := int32(0)
+	for i, proof := range p.ShareToRowRootProof {
+		nmtProof := nmt.NewInclusionProof(
+			int(proof.Start),
+			int(proof.End),
+			proof.Nodes,
+			true,
+		)
+		sharesUsed := proof.End - proof.Start
```
[the remark this answers is not in the document]
Comment (rach-id) [reply blob/blob.go]: [28cb03f](https://github.com/celestiaorg/celestia-node/pull/3610/commits/28cb03f036a42f938a2eff43d0886176a9575e78)

File: blob/blob.go [hunk 27]
```diff
@@ -16,42 +19,70 @@ import (
 	"github.com/celestiaorg/celestia-node/share"
 )
 
+//nolint:unused
```
Comment (Wondertan) [inline blob/blob.go:22]: Weird that its still here, even there was a comment addressing it

Comment (Wondertan) [inline blob/blob.go:22]: Weird that its still here, even there was a comment addressing it

File: blob/service.go [hunk 28]
```diff
@@ -347,127 +353,193 @@ func (s *Service) retrieve(
...
-		rowIndex++
-		if sharesParser.isEmpty() {
-			proofs = nil
-		}
-	}
+				// create the share to row root proofs
+				shareToRowRootProofs, _, err := pkgproof.CreateShareToRowRootProofs(
+					squareSize,
+					//nolint:lll
+					edsShares[inclusiveBlobStartRowIndex-inclusiveNamespaceStartRowIndex:exclusiveBlobEndRowIndex-inclusiveNamespaceStartRowIndex],
```
Comment (cristaloleg) [inline blob/service.go:514]: Is there a bug? lower bound has 2 vars with `inclusive` and upped bound has 1 var with `exclusive` and 1 with `inclusive`. Looks sus.

Can we split this into 3 lines, please?
```
start := inclusiveBlobStartRowIndex-inclusiveNamespaceStartRowIndex
end := exclusiveBlobEndRowIndex-inclusiveNamespaceStartRowIndex
...
edsShares[start : end]
```

Comment (cristaloleg) [reply blob/service.go:514]: Also not symmetrical `...StartRow - ...StartRow` BUT `...EndRow - ...StartRow`. Which might be ok but sus 🤷

File: blob/blob.go (hunk 27 above)
Comment (rach-id) [reply blob/blob.go:22]: https://github.com/celestiaorg/celestia-node/pull/3610#discussion_r1700202243

File: blob/service.go [hunk 29]
```diff
@@ -347,127 +353,193 @@ func (s *Service) retrieve(
...
+	exclusiveNamespaceEndRowIndex := inclusiveNamespaceStartRowIndex
+	for _, row := range header.DAH.RowRoots[inclusiveNamespaceStartRowIndex:] {
+		if namespace.IsOutsideRange(row, row) {
+			break
+		}
+		exclusiveNamespaceEndRowIndex++
+	}
+	if exclusiveNamespaceEndRowIndex == inclusiveNamespaceStartRowIndex {
+		return nil, nil, fmt.Errorf("couldn't find the row index of the namespace end")
+	}
```
Comment (cristaloleg) [inline blob/service.go:386]: Q: so, it's ok when we just iterate over `header.DAH.RowRoots[inclusiveNamespaceStartRowIndex:]` till the end. This means that we cover all ranges, am I right?

File: blob/blob.go (hunk 27 above)
Comment (Wondertan) [reply blob/blob.go:22]: That comment referred to a diff code and path tho

Comment (Wondertan) [reply blob/blob.go:22]: Although, it doesn't hurt to keep it if we keep some other deadcode anyway

Comment (rach-id) [reply blob/blob.go:22]: > That comment referred to a diff code and path tho

yes, but that's the only code using this. So the linter also detects this as unused.

File: blob/service.go (hunk 29 above)
Comment (rach-id) [reply blob/service.go:386]: yes, the namespaces are ordered

File: blob/service_test.go [hunk 30]
```diff
@@ -1061,3 +1061,124 @@ func generateCommitmentProofFromBlock(
...
+			}(),
+			blob: func() Blob {
+				b := *blob
+				b.Commitment = []byte{0x1}
+				return b
+			}(),
+			expectErr: true,
+		},
+		{
+			name:     "mismatched number of share proofs and row proofs",
```
Comment (renaynay) [inline blob/service_test.go:1148]: this is same test case as invalid proof

**Review by renaynay — COMMENTED** [review]
one comment here otherwise thank u for tests :)

File: blob/service_test.go (hunk 30 above)
Comment (rach-id) [reply blob/service_test.go:1149]: [857e3ed](https://github.com/celestiaorg/celestia-node/pull/3610/commits/857e3ed106e17a43ca88eff68fc2672ca1fddb7c)

**Comment by vgonkivs** [comment]
Btw, unit test failed

File: blob/service.go (hunk 28 above)
Comment (cristaloleg) [reply blob/service.go]: No bug in indexes, right? Just asking for the confirmation.

**Comment by rach-id** [comment]
Unit test fixed: [712c272](https://github.com/celestiaorg/celestia-node/pull/3610/commits/712c2728ea118f571019133bb09e28467a29d3d4)

was a good catch 👍

**Review by vgonkivs — APPROVED** [review]

File: blob/service.go (hunk 28 above)
Comment (rach-id) [reply blob/service.go]: yes, no bug. We're shifting the indexes by `inclusiveNamespaceStartRowIndex` because the EDS shares start from `[inclusiveNamespaceStartRowIndex: exclusiveNamespaceEndRowIndex]`

File: blob/blob.go (hunk 22 above)
Comment (renaynay) [reply blob/blob.go:42]: Yes bc this error doesn't have anything to do with the proof, just the blob

**Review by renaynay — APPROVED** [review]
Thank you @rach-id let's get this over the line :)

File: blob/blob.go [hunk 31]
```diff
@@ -16,42 +19,70 @@ import (
...
-
-func (p Proof) Len() int { return len(p) }
+// Proof constructs the proof of a blob to the data root.
+type Proof struct {
+	// ShareToRowRootProof the proofs of the shares to the row roots they belong to.
+	// If the blob spans across multiple rows, then this will contain multiple proofs.
+	ShareToRowRootProof []*tmproto.NMTProof
+	// RowToDataRootProof the proofs of the row roots containing the blob shares
+	// to the data root.
+	RowToDataRootProof coretypes.RowProof
```
Comment (Wondertan) [inline blob/blob.go:32]: Q: why do we call DataRoot if inside header it is called `DataHash`?

Comment (rach-id) [reply blob/blob.go:32]: That's an existential question. From the dawn of existence, the name `DataRoot` has echoed through the ages, a timeless whisper woven into the fabric of eternity.

That's a theory vs implementation issue... In the fraud proofs paper, it's referred to as `DataRoot`. In tendermint, it's called `DataHash`, and AFAIK we never really discussed choosing one over the other.

File: blob/blob.go (hunk 22 above)
Comment (rach-id) [reply blob/blob.go:42]: [3630f33](https://github.com/celestiaorg/celestia-node/pull/3610/commits/3630f332e858c72ae0445e1abb9871b0fcc1e946)
